src/Raven.Server/Documents/DocumentDatabase.cs

```diff
@@ -753,21 +765,49 @@ private void OnClusterTransactionCompletion(ClusterTransactionCommand.SingleClus
             {
                 var index = command.Index;
                 var options = mergedCommands.Options[index];
-                Task indexTask = null;
+                var taskResult = new ClusterTransactionResult
+                {
+                    GeneratedResult = mergedCommands.Replies[index],
+                };
                 if (options.WaitForIndexesTimeout != null)
                 {
-                    indexTask = BatchHandlerProcessorForBulkDocs.WaitForIndexesAsync(this, options.WaitForIndexesTimeout.Value,
+                    var indexTask = BatchHandlerProcessorForBulkDocs.WaitForIndexesAsync(this, options.WaitForIndexesTimeout.Value,
                         options.SpecifiedIndexesQueryString, options.WaitForIndexThrow,
                             mergedCommands.LastDocumentEtag, mergedCommands.LastTombstoneEtag, mergedCommands.ModifiedCollections);
-                }
 
-                var result = new ClusterTransactionCompletionResult
+                    var removeTask = ServerStore.Cluster.ClusterTransactionWaiter.CreateTask(command.Options.TaskId);
+
+                    /*
+                        the remove task that we are creating here is relevant only for a failover with wait for index:
+                        We need the task only when we do a failover during cluster transaction (batch handler) with 'wait for index'.
+                        Only then we will want the task that created here, because we'll want to 'wait for index' in the new node after the failover.
+                    */
+
+                    indexTask.ContinueWith(t =>
+                    {
+                        try
+                        {
+                            t.GetAwaiter().GetResult(); // Task t is already completed here
+                            RachisLogIndexNotifications.NotifyListenersAbout(index, null);
+                            ServerStore.Cluster.ClusterTransactionWaiter.SetResult(options.TaskId, taskResult);
+                        }
+                        catch (Exception e)
+                        {
+                            RachisLogIndexNotifications.NotifyListenersAbout(index, e);
+                            ServerStore.Cluster.ClusterTransactionWaiter.SetException(options.TaskId, e);
+                        }
```

Comment:
Weren't we already calling that ealier?

Reply:
no, we set the task only in the end of the indexing or (xor) in its completion in the database (if we don't need to wait form indexing).
It happens only once.

Reply:
It happens when the cluster transaction raft command is being committed only when the database is shared (that was the behavior before my changes).